File: kora-app-annotation-processor/src/test/java/ru/tinkoff/kora/kora/app/annotation/processor/app/AppWithFactories2.java
package ru.tinkoff.kora.kora.app.annotation.processor.app;

import ru.tinkoff.kora.application.graph.TypeRef;
import ru.tinkoff.kora.common.KoraApp;
import ru.tinkoff.kora.common.annotation.Root;

import java.util.List;

@KoraApp
public interface AppWithFactories2 {
    @Root
    default Class1 class1(GenericClass<List<Class1>, String> dependency) {
        return new Class1();
    }

    default <T> GenericClassImpl<List<T>> factory2(TypeRef<T> typeRef) {
        return new GenericClassImpl<>();
    }

    class GenericClass<T, Q> {}

    class GenericClassImpl<T> extends GenericClass<T, String> {}

    class Class1 {}
}

File: kora-app-symbol-processor/src/test/kotlin/ru/tinkoff/kora/kora/app/ksp/app/AppWithFactories2.kt
package ru.tinkoff.kora.kora.app.ksp.app

import ru.tinkoff.kora.application.graph.TypeRef
import ru.tinkoff.kora.common.KoraApp
import ru.tinkoff.kora.common.annotation.Root

@KoraApp
interface AppWithFactories2 {
    @Root
    fun class1(dependency: GenericClass<List<Class1>, String>): Class1 {
        return Class1()
    }

    @Root
    fun <T> factory2(typeRef: TypeRef<T>): GenericClassImpl<List<T>> {
        return GenericClassImpl()
    }

    open class GenericClass<T, Q>
    class GenericClassImpl<T> : GenericClass<T, String>()
    class Class1
}
